lib/licensed/sources/nuget.rb

```diff
@@ -0,0 +1,226 @@
+# frozen_string_literal: true
+require "json"
+
+module Licensed
+  module Sources
+    # Only supports ProjectReference (project.assets.json) style restore used in .NET Core.
+    # Does not currently support packages.config style restore.
+    class NuGet < Source
+      def self.type
+        "nuget"
+      end
+
+      # Match nuspec license expressions like <license type="expression">MIT</license>
+      class NuspecLicenseExpressionMatcher < Licensee::Matchers::Package
+        LICENSE_EXPRESSION_REGEX = /
+          <license\s*type\s*=\s*\"\s*expression\s*\"\s*>\s*([a-z\-0-9\.]+)\s*<\/license>
+        /ix.freeze
+
+        def license_property
+          return @license_property if defined?(@license_property)
+          match = @file.content.match LICENSE_EXPRESSION_REGEX
+          if match && match[1]
+            @license_property = match[1].downcase
+          end
+        end
+      end
+
+      class NuspecFile < Licensee::ProjectFiles::ProjectFile
+        def possible_matchers
+          [NuspecLicenseExpressionMatcher]
+        end
+      end
+
+      # A pseudo-license file where the license is already known (e.g. inferred from a URL)
+      class KnownLicenseFile < Licensee::ProjectFiles::ProjectFile
+        def initialize(license, metadata)
+          super(license.content, metadata)
+          @license = license
+        end
+
+        def license
+          return @license
+        end
+
+        def confidence
+          100
+        end
+      end
+
+      class NuGetDependency < Licensed::Dependency
+        LICENSE_FILE_REGEX = /<license\s*type\s*=\s*\"\s*file\s*\"\s*>\s*(.*)\s*<\/license>/ix.freeze
+        LICENSE_URL_REGEX = /<licenseUrl>\s*(.*)\s*<\/licenseUrl>/ix.freeze
+
+        def license
+          # By default, multiple of the same/similar licenses (e.g. a package with LICENSE.txt, license, and licenseUrl)
+          # will incorrectly result in an "other" license.
+
+          # Treat license expressions as most trustworthy, then local licenses, then remote licenses, then any license files
+          return @license if defined? @license
+          return @license = package_file.license if package_file && package_file.license
+          return @license = nuspec_local_license_file.license if nuspec_local_license_file && nuspec_local_license_file.license
+          return @license = nuspec_remote_license_file.license if nuspec_remote_license_file && nuspec_remote_license_file.license
```

Comment:
Thoughts on fetching `licenseUrl` if other licenses were already found from a file or license expression?

On one hand, it seems wise to avoid doing so, on the other hand the online license may be different than the file.

Reply:
I no longer override `license`, but I still have custom logic in `project_files` to only download licenseUrl if neither licensee nor the local file search found a recognized license.

With the change I did in `licensee`, it's no longer possible to differentiate if `licensee` actually matched on `<license>` or `<licenseUrl>`. I guess I could make another change in licensee to add that back in, but to be honest I'd rather not if not necessary.

Thoughts?

Reply:
That custom logic sounds right to me.